Automattic/wp-calypso
# Calypso apps: build meta json file with cache buster

#### Changes proposed in this Pull Request

This PR generates a "build meta" JSON file during Webpack for each of the JS Calypso apps. This is very similar to how the `cache_buster.txt` file was generated for Notifications. (Indeed, this new file replaces cache_buster.txt.) It basically just populates a JSON file with some basic git information, using `git describe...` for the cache buster value. (The same format as was used for notifications.)

We were also generating a `cache_buster.txt` file in TeamCity, and this new JSON file also replaces that. I've updated the TeamCity builds to utilize the JSON file when it exists. (TeamCity will also add a `build_number` property to this JSON file for extra context.)

This works by creating a small wrapper around HTMLWebpackPlugin, and then using a JSON template file. (This plugin was placed in `@automattic/calypso-apps-builder`, but the location is arbitrary.) Each of the webpack configs for JS Calypso apps then add this plugin, and the build_meta file is automatically placed in their dist directories.

Once this PR is merged:
- notifications, wpcom-block-editor, happychat, and inline-help will contain `build_meta.json` in their release archives. They will no longer include `build_meta.txt`
- ETK and o2-blocks still use `build_meta.txt`.
- TeamCity will switch to use whichever file exists, which maintains back compat for the few commits which might need it.

#### Testing instructions
- Inspect the archives for each of the calypso apps mentioned above, and ensure they have a `build_meta.json` file.
- Test D76214-code with these new apps and ensure cache busters are updated correctly.
